app/commands/track/update_build_status.rb

```diff
@@ -0,0 +1,368 @@
+      num_concepts_target:,
+      created: taught_concepts.map { |concept| serialize_concept(concept) }
+    }
+  end
+
+  def num_concepts_target
+    NUM_CONCEPTS_TARGETS.find { |target| taught_concepts.size < target } || taught_concepts.size
+  end
+
+  memoize
+  def taught_concepts = Concept.where(id: concept_taught_exercise.keys).to_a
+
+  memoize
+  def concept_taught_exercise
+    Exercise::TaughtConcept.
+      where(exercise: active_concept_exercises).
+      pluck(:track_concept_id, :exercise_id).
+      map { |concept_id, exercise_id| { concept_id => exercise_id } }.
+      reduce({}, :merge)
+  end
+
+  def concept_exercises
+    {
+      num_exercises: active_concept_exercises.size,
+      num_exercises_target: concept_exercises_num_exercises_target,
+      created: active_concept_exercises.map { |exercise| serialize_exercise(exercise) }
+    }
+  end
+
+  def concept_exercises_num_exercises_target
+    NUM_CONCEPT_EXERCISES_TARGETS.find { |target| active_concept_exercises.size < target } || active_concept_exercises.size
+  end
+
+  def practice_exercises
+    {
+      num_exercises: active_practice_exercises.size,
+      num_exercises_target: practice_exercises_num_exercises_target,
+      created: active_practice_exercises.map { |exercise| serialize_exercise(exercise) },
+      volunteers: practice_exercises_volunteers,
+      health: practice_exercises_health
+    }
+  end
+
+  def practice_exercises_num_exercises_target
+    num_unimplemented = Track::UnimplementedPracticeExercises.(track.reload).size
+    max_target = track.num_exercises + num_unimplemented
+
+    NUM_PRACTICE_EXERCISES_TARGETS.find { |target| active_practice_exercises.size < target } || max_target
+  end
+
+  memoize
+  def num_unimplemented_practice_exercises
+    Track::UnimplementedPracticeExercises.(track).size
+  end
+
+  def practice_exercises_volunteers
+    authors = User.with_attached_avatar.where(id:
+      Exercise::Authorship.where(exercise: active_practice_exercises).select(:user_id))
+    contributors = User.with_attached_avatar.where(id:
+      Exercise::Contributorship.where(exercise: active_practice_exercises).select(:user_id))
+
+    serialize_volunteers(authors, contributors)
+  end
+
+  memoize
+  def practice_exercises_health
+    return :missing if active_practice_exercises.empty?
+    return :exemplar if active_practice_exercises.size >= NUM_PRACTICE_EXERCISES_TARGETS.last
+    return :needs_attention if active_practice_exercises.size < NUM_PRACTICE_EXERCISES_TARGETS.first
+
+    :healthy
+  end
+
+  memoize
+  def active_concept_exercises = track.concept_exercises.where(status: %i[active beta]).order(:position).to_a
+
+  memoize
+  def active_practice_exercises = track.practice_exercises.where(status: %i[active beta]).order(:position).to_a
+
+  def average_number_per_day(query, model)
+    total_count = query.where("#{model.table_name}.created_at >= ?", Time.current - NUM_DAYS_FOR_AVERAGE.days).count
```

Comment:
Iterating by date is slow. This may be quicker if you do something like:

```
first_id = query.where("#{model.table_name}.created_at >= ?", Time.current - NUM_DAYS_FOR_AVERAGE.days).order(:id).pick(:id)
total_count = query.where("#{model.table_name}.id >= ?", first_id).count
```

But it might also not be, as SQL might not be able to optimise it enough. But I think it will be.

Reply:
Cool method though.